Bug 1986453: Check for API server and node versions skew

QiWang19 opened the pull request:
ref: https://issues.redhat.com/browse/OCPNODE-595
replace https://github.com/openshift/machine-config-operator/pull/2552

**- What I did**

Check for API server and node versions skew.
Update with the message the Kube API version is skew too far, but do not force
`Upgradeable=False` according to enhancement https://github.com/openshift/enhancements/pull/762/files

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

QiWang19 commented:
/assign @sinnykumari

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
 	return optr.updateStatus(co, coStatus)
 }
 
+func (optr *Operator) isKubeletSkewTooFar() (bool, error) {
```
```suggestion
func (optr *Operator) isKubeletSkewSupported() (bool, error) {
```

kikisdeliveryservice replied:
also please add func description above declaration

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
+	// don't overwrite status if we didn't run
+	if ran, err := optr.isKubeletSkewTooFar(); ran && err != nil {
+		if coStatus.Status != configv1.ConditionFalse {
+			coStatus.Reason = "KubeletSkewTooFar"
```
```suggestion
			coStatus.Reason = "KubeletSkewUnsupported"
```

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
+		return true, err
+	}
+
+	skewedNodeErrors := []error{}
```
do we really want to have a list of errors and print a list at the CO level?

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
 	}
 
+	// don't overwrite status if we didn't run
+	if ran, err := optr.isKubeletSkewTooFar(); ran && err != nil {
```
why ran when you aren't checking if it ran but if the skew is supported?

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
+			skewedNodeErrors = append(skewedNodeErrors, fmt.Errorf("node/%v must be updated or removed before the cluster can upgrade, current version %v", node.Name, kubeletVersion))
+		}
+	}
+	return true, errors.NewAggregate(skewedNodeErrors)
```
I'm kind of confused about what the bool return values in this func signify? is it t/f skew supported? or t/r function ran?  I think it should be the former, but it's not clear reading this.

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
+			skewedNodeErrors = append(skewedNodeErrors, fmt.Errorf("node/%v has malformed kubelet version %q: %v", node.Name, kubeletVersion, err))
+		}
+		if nodeMinorVersion+2 <= kubeAPIServerMinorVersion {
+			skewedNodeErrors = append(skewedNodeErrors, fmt.Errorf("node/%v must be updated or removed before the cluster can upgrade, current version %v", node.Name, kubeletVersion))
```
I think this message should be more explicit, ie tell them how many versions they are off/what the skew is

kikisdeliveryservice reviewed (commented):
some initial questions/comments

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
+	if ran, err := optr.isKubeletSkewTooFar(); ran && err != nil {
+		if coStatus.Status != configv1.ConditionFalse {
+			coStatus.Reason = "KubeletSkewTooFar"
+			coStatus.Message = err.Error()
```
Feel like this really should be a const Message

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
+			coStatus.Reason = "KubeletSkewTooFar"
+			coStatus.Message = err.Error()
+		} else if err != nil {
+			coStatus.Reason = coStatus.Reason + "::" + "KubeletSkewTooFar"
```
what are we doing Reason::KubeletSkewTooFar?instead of just KubeletSkewTooFar?

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
+			coStatus.Message = err.Error()
+		} else if err != nil {
+			coStatus.Reason = coStatus.Reason + "::" + "KubeletSkewTooFar"
+			coStatus.Message = coStatus.Message + "\n" + err.Error()
```
similar Q: coStats.Message+error?

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
 		coStatus.Message = "One or more machine config pools are updating, please see `oc get mcp` for further details"
 	}
 
+	// don't overwrite status if we didn't run
```
I'm not convinced we should be reporting on a node level. In a large cluster this could  result  in a big list of nodes when 1) the MCO doesn't think via individual nodes but rather pools and 2) a long list is unmangeable as a ui

I think this should probably be reported at a pool level.

kikisdeliveryservice replied:
a message like: Worker pool contains nodes with an unsupported kubelet skew. Please see `oc get nodes` for details and ensure all nodes have a kubelet Version of at least $version.

Would be more concise but also more descriptive.

QiWang19 replied:
@kikisdeliveryservice  What is reporting at the pool level? The version skew should be checked for each node, follow the same implementation as https://github.com/openshift/machine-config-operator/blob/db56bd2856b4892f3b72e6056904a29676d362d4/pkg/operator/status.go#L289, the message is reported at pool level.  Could you elaborate?

kikisdeliveryservice replied:
I have a suggestion above: https://github.com/openshift/machine-config-operator/pull/2658/files#r663253475

kikisdeliveryservice reviewed (commented):
a few more comments

/hold

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
+		}
+		nodeMinorVersion, err := getMinorVersion(kubeletVersion)
+		if err != nil {
+			skewedNodeErrors = append(skewedNodeErrors, fmt.Errorf("node/%v has malformed kubelet version %q: %v", node.Name, kubeletVersion, err))
```
what does this mean to a user? what are they supposed to do to fix?

QiWang19 replied:
This only provides the user with more information that there's an error when checking the version. Might be helpful debugging if they run into other issues but no way to fix it at this point. Should we `glog.Error(err)` this error or silently ignore it?

kikisdeliveryservice replied:
Would this error manifest as another error that already exists? IE how does a user get into this situation and would it already be reflected in the state of the pool (would the pool be degraded, etc..)

QiWang19 replied:
> Would this error manifest as another error that already exists? IE how does a user get into this situation and would it already be reflected in the state of the pool (would the pool be degraded, etc..),would it fail the kubelet health check, etc..

I agree with the above situation. I tend to only report the message and reason if the version has been successfully checked and compared to verify there's a version skew.

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
+		if coStatus.Status != configv1.ConditionFalse {
+			coStatus.Reason = "KubeletSkewTooFar"
+			coStatus.Message = err.Error()
+		} else if err != nil {
```
aren't we already guaranteed err!=nil from `ran && err != nil` above?

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +293,82 @@ func (optr *Operator) syncUpgradeableStatus() error {
+	return true, errors.NewAggregate(skewedNodeErrors)
+}
+
+func getMinorVersion(version string) (int, error) {
```
```suggestion
func getMinorKubeletVersion(version string) (int, error) {
```

kikisdeliveryservice commented:
@QiWang19 I know that this PR is based on someone else's previous PR, but I think we can refine it. What most of my comments are asking for is: using that work as a baseline, how can we make this into an understandable & actionable status for a user. Also keep in mind a pool can be large, so the info needs to be easy to consume if we have, say, 50 nodes that have unsupported skew.  

Users can get into this situation bc a pool is paused and so it did not upgrade to match the apiserver and is still at an older version. They will likely have to let those pools upgrade (to at a minimum a supported skew) before they can initiate another clusterwide upgrade (which would cause the kubeapiserver to get even further away).  So we need to think about telling them the state in a meaningful way, but also give them some hint about what they need to do to remedy it.

Happy to discuss further if you have any questions. =)

QiWang19 commented:
@kikisdeliveryservice Thanks for the explanation. I have cleaned up some reviews. PTAL.

QiWang19 commented:
/retest

QiWang19 commented:
/retest

QiWang19 commented:
/retest

QiWang19 commented:
@kikisdeliveryservice Thanks for the explanation. I have cleaned up some reviews. Could you PTAL?

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,79 @@ func (optr *Operator) syncUpgradeableStatus() error {
+	// don't overwrite status if we didn't run
+	if supported, v := optr.isKubeletSkewSupported(); !supported {
+		coStatus.Reason = "KubeletSkewUnsupported"
+		coStatus.Message = fmt.Sprintf("machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least %v.", v)
```
`upgrade all nodes so that they have a kubelet Version of at least %v.", v`

If v is the apiserver version, the nodes don't have to be updated to that version but to that version-2 at a minimum no?

QiWang19 replied:
> `upgrade all nodes so that they have a kubelet Version of at least %v.", v`
> 
> If v is the apiserver version, the nodes don't have to be updated to that version but to that version-2 at a minimum no?

The version skew does not support version - 2 version number. It should be the version with n-1 (n is the minor version of the version). If we want to report the version with a specific version number, will need another parsing function to do the calculation. https://github.com/openshift/machine-config-operator/pull/2658/files#diff-8a6a064b4db19eb3b29e5280447ac3ba9496e847391489d25c1282767c3354c5R371
Do you think it worth adding this func?

kikisdeliveryservice replied:
I think your update covers it by printing minimalsupported version. :+1:

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,79 @@ func (optr *Operator) syncUpgradeableStatus() error {
 
+	// don't overwrite status if we didn't run
+	if supported, v := optr.isKubeletSkewSupported(); !supported {
+		coStatus.Reason = "KubeletSkewUnsupported"
```
As is this would be overwriting a degraded and updating status, I think we probably only want to do this if we are available but not updating/degraded? Doesn't seem like this should preempt those more impt messages? wdyt

kikisdeliveryservice reviewed (commented):
a few more comments

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,99 @@ func (optr *Operator) syncUpgradeableStatus() error {
+		if supported, v := optr.isKubeletSkewSupported(); !supported {
+			coStatus.Reason = "KubeletSkewUnsupported"
+			coStatus.Message = fmt.Sprintf("machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least %v.", getMinimalSkewSupportNodeVersion(v))
+		}
```
can we also add an event here?

see an example here: https://github.com/openshift/machine-config-operator/blob/4f555d05550bca9a67f5354afa2721cea5e36b02/pkg/operator/status.go#L44

QiWang19 replied:
@kikisdeliveryservice I added the event. PTAL.

sinnykumari replied:
Shouldn't this event be of type "EventTypeWarning" ?

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,99 @@ func (optr *Operator) syncUpgradeableStatus() error {
+			glog.Error(err)
+			continue
+		}
+		if nodeMinorVersion+2 <= kubeAPIServerMinorVersion {
```
note: we're checking on this

QiWang19 replied:
@kikisdeliveryservice  I checked this with @rphillips . We can keep it as is. Considering the 2nd example here https://docs.google.com/document/d/1yXzWPBWmKLuKBkvOKMk0FusyBpUqjyT3hGXYL8tUafU/edit#. we get into this situation when we do upgrade kube-apiserver is at 1.21 and 1.20 , then the kubelet can only support 1.20 and 1.19.

kikisdeliveryservice replied:
Wouldn't <= mean that we only support 1? if apiserver is at 1.20 and node is at 1.18 and I add 2 it's = apiserver and not supported. Is this correct?

deads2k replied:
Up to n-2 is support.

1. 1.21 kas supports 1.21, 1.20, 1.19 kubelet
2. 1.20 kas supports 1.20, 1.19, 1.18 kubelet

The allowable skew is the intersection, in this case 1.20 and 1.19.

QiWang19 replied:
> Up to n-2 is support.
> 
>     1. 1.21 kas supports 1.21, 1.20, 1.19 kubelet
> 
>     2. 1.20 kas supports 1.20, 1.19, 1.18 kubelet
> 
> 
> The allowable skew is the intersection, in this case 1.20 and 1.19.

@deads2k So the skew calculation should take into account the upgrade, like both 1.21 and 1.20 in the above example?

kikisdeliveryservice replied:
right but my question is that <= correct?

and isn't apiserver has various versions only applicable when we're mid-update? post update, they would be the same.

kikisdeliveryservice commented:
Just noting that when this is done, we need to get the commits updated so Qi is a co-author.

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,99 @@ func (optr *Operator) syncUpgradeableStatus() error {
+	if !updating && !degraded {
+		if supported, v := optr.isKubeletSkewSupported(); !supported {
+			coStatus.Reason = "KubeletSkewUnsupported"
+			coStatus.Message = fmt.Sprintf("machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least %v.", getMinimalSkewSupportNodeVersion(v))
```
There's been a bunch of great changes on this PR, so I'm giving this message a little thought.

QiWang19 commented:
/retest

QiWang19 commented:
/retest

kikisdeliveryservice commented:
Would like an answer on my question about:

nodeMinorVersion+2 <= kubeAPIServerMinorVersion

means that if i have a 1.19 and apiserver is 1.21 we are saying this isn't supported and they must upgrade even tho the skew is a supported 2 versions.  shouldn't this be:

nodeMinorVersion+2 < kubeAPIServerMinorVersion

rphillips commented:
@kikisdeliveryservice According to this [doc](https://docs.google.com/presentation/d/1PJvN6-hLm9Vr8KuecMmiyk9J5vszGDcTAsppKLUet70/edit#slide=id.gbab5de315b_0_112) the version skew will always be the current version and the next version. The n+2 version should be flagged as an error.

kikisdeliveryservice commented:
>the version skew will always be the current version and the next version. The n+2 version should be flagged as an error.

@rphillips I'm really confused on this bc I thought the whole point was to allow an n+2 version skew, which is also what upstream is?

kikisdeliveryservice commented:
but also the file that you are linking is allowing n+2

kikisdeliveryservice commented:
Wait I think my issue is the phrasing bc n+2 is supported, but we're _advising_ you at that point to upgrade if you are not intending to stay there.. I need more coffee.

rphillips commented:
| Openshift Version | Kubelet Version |
| ------------------------| ----------------------|
| 4.6 | 1.0 |
| 4.7 | 1.1 |
| 4.8 | *1.2* |
| 4.9 | 1.3 |
| 4.10 | *1.4* |

Control Plane upgrades to 4.9.x  & Worker nodes remain on 4.8.x, then the control plane upgrades to 4.10. Kirsten is correct... It shouldn't be <=.

rphillips commented:
```diff
diff --git a/pkg/operator/status.go b/pkg/operator/status.go
index 62f0ea81..3bf5b06f 100644
--- a/pkg/operator/status.go
+++ b/pkg/operator/status.go
@@ -346,7 +346,7 @@ func (optr *Operator) isKubeletSkewSupported() (skewSupported bool, kubeAPIServe
                        glog.Error(err)
                        continue
                }
-               if nodeMinorVersion+2 <= kubeAPIServerMinorVersion {
+               if nodeMinorVersion+2 < kubeAPIServerMinorVersion {
                        skewSupported = false
                        break
                }
diff --git a/pkg/operator/status_test.go b/pkg/operator/status_test.go
index 83cd8b92..15209418 100644
--- a/pkg/operator/status_test.go
+++ b/pkg/operator/status_test.go
@@ -637,7 +637,7 @@ func TestKubeletSkewSupported(t *testing.T) {
                ObjectMeta: metav1.ObjectMeta{Name: "kube-apiserver"},
                Status: configv1.ClusterOperatorStatus{
                        Versions: []configv1.OperandVersion{
-                               {Name: "kube-apiserver", Version: "1.20"},
+                               {Name: "kube-apiserver", Version: "1.21"},
                        },
                },
        }
@@ -705,7 +705,7 @@ func TestKubeletSkewSupported(t *testing.T) {
        if upgradeable.Status != configv1.ConditionTrue {
                t.Fatal(upgradeable)
        }
-       if upgradeable.Message != "machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least 1.19." {
+       if upgradeable.Message != "machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least 1.20." {
                t.Fatal(upgradeable)
        }
        if upgradeable.Reason != "KubeletSkewUnsupported" {
```

QiWang19 commented:
> ```diff
> -       if upgradeable.Message != "machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least 1.19." {
> +       if upgradeable.Message != "machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least 1.20." {
> ```

@rphillips should this at least version be 1.19? The test case kube-apiserver 1.21, node version 1.18. If the policy allows at most 2 versions skew. The minimum version they need to upgrade to is 1.19.

rphillips commented:
@QiWang19 Agreed... Can you fix that up?

QiWang19 commented:
> @QiWang19 Agreed... Can you fix that up?

Fixed.

rphillips commented:
/lgtm

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,100 @@ func (optr *Operator) syncUpgradeableStatus() error {
+	if !updating && !degraded {
+		if supported, v := optr.isKubeletSkewSupported(); !supported {
+			coStatus.Reason = "KubeletSkewUnsupported"
+			coStatus.Message = fmt.Sprintf("machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least %v.", getMinimalSkewSupportNodeVersion(v))
```
```suggestion
			coStatus.Message = fmt.Sprintf("one or more nodes have an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least %v.", getMinimalSkewSupportNodeVersion(v))
```

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,100 @@ func (optr *Operator) syncUpgradeableStatus() error {
+	return int(minorVersion), nil
+}
+
+// getMinimalSkewSupportNodeVersion returns the minimal node kubelet version should be
```
```suggestion
// getMinimalSkewSupportNodeVersion returns the minimal supported node kubelet version.
```

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,100 @@ func (optr *Operator) syncUpgradeableStatus() error {
+}
+
+// getMinimalSkewSupportNodeVersion returns the minimal node kubelet version should be
+// upgrade to for skew support.
```
```suggestion
```

rphillips commented:
/test e2e-aws-upgrade-paused

rphillips commented:
/test openshift-upgrade-aws-pausedworkers

sinnykumari commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,100 @@ func (optr *Operator) syncUpgradeableStatus() error {
+		if supported, v := optr.isKubeletSkewSupported(); !supported {
+			coStatus.Reason = "KubeletSkewUnsupported"
+			coStatus.Message = fmt.Sprintf("machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least %v.", getMinimalSkewSupportNodeVersion(v))
+			optr.eventRecorder.Eventf(co, corev1.EventTypeNormal, "KubeletSkewUnsupported", fmt.Sprintf("machine config pool contains nodes with an unsupported kubelet version skew with kube-apiserver version %v.", v))
```
Shouldn't this event be of type "EventTypeWarning"?

kikisdeliveryservice replied:
agree good catch

sinnykumari commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,100 @@ func (optr *Operator) syncUpgradeableStatus() error {
 }
 
+// isKubeletSkewSupported checks the version skew of kube-apiserver and node kubelet version.
+// Returns false is the version skew is not supported. version skew > 2 is not supported.
```
nit: rewording "Returns false if the version skew is not supported"

sinnykumari commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,100 @@ func (optr *Operator) syncUpgradeableStatus() error {
+	skewSupported = true
+	nodes, err := optr.nodeLister.List(labels.Everything())
+	if err != nil {
+		return skewSupported, ""
```
I am confused, in case of error shouldn't we returning "skewSupported=false"?

sinnykumari replied:
Possibly also return error and set error message when an error has been encountered.

QiWang19 replied:
See comment https://github.com/openshift/machine-config-operator/pull/2658#discussion_r663273670. The former discussion has question about this kind of error should cause failure in other processes and the message returned from here would not be very useful for customers. So ignore the error and the skew hasn't been checked at this point, shouldn't return false.

kikisdeliveryservice replied:
I think what @sinnykumari is pointing out is that the issue is that if there is an error we're saying the skew IS supported, when we don't know?

kikisdeliveryservice replied:
We use this func here:
```
if supported, v := optr.isKubeletSkewSupported(); !supported 
```
but if the kubeAPIServerVersion="" is intended to indicate that we didn't actually check the skew, you'd have to look at the value of v there.

QiWang19 replied:
> but if the kubeAPIServerVersion="" is intended to indicate that we didn't actually check the skew, you'd have to look at the value of v there.

Here it failed to get the node kublet version. Could you explain your idea?

kikisdeliveryservice replied:
Thinking about sinny's comment we can either:
1) log in the operator logs the error and the node
and also maybe:
2) log in the status the failure to check the skew and to see the operator logs for details

@sinnykumari wdyt?

sinnykumari replied:
Definitely we should do 1, also doing 2 makes sense i.e. reporting error message to MCO coStatus with error message and details on where to look for further debugging.

QiWang19 replied:
> Definitely we should do 1, also doing 2 makes sense i.e. reporting error message to MCO coStatus with error message and details on where to look for further debugging.

@sinnykumari  Just to be clear, is this correct: 
- log in the operator logs is using glog.Error()
- log in the status is the coStatus.Message

QiWang19 replied:
@sinnykumari PTAL. Added the error to coStatus and glog.Info() to log the current status.

rphillips replied:
This should now be resolved in the latest commit.

sinnykumari commented on pkg/operator/status_test.go:
```diff
@@ -621,3 +631,84 @@ func TestInClusterBringUpStayOnErr(t *testing.T) {
 	assert.False(t, optr.inClusterBringup)
 }
+
+func TestKubeletSkewSupported(t *testing.T) {
```
Right now, we have test for happy path.To make it more robust, it feels like we should add more coverage in this unit tests by adding more test cases like kubelet version is something like  v1.21.0-rc.0+6143dea, when kubelet skew is supported, etc.

QiWang19 replied:
@sinnykumari Updated the test for both path tests and added more version parsing tests.

wking commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,105 @@ func (optr *Operator) syncUpgradeableStatus() error {
+	if !updating && !degraded {
+		if supported, v := optr.isKubeletSkewSupported(); !supported {
+			coStatus.Reason = "KubeletSkewUnsupported"
+			coStatus.Message = fmt.Sprintf("one or more nodes have an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least %v.", getMinimalSkewSupportNodeVersion(v))
```
I'm just noticing that [the enhancement][1] doesn't currently have much detail here.  There's some follow-up work still in flight [here][2] with:

> The Machine Config Operator should report a message in its reporting that admins should perform an action.

But based on my fuzzy recollection of some in-person meeting at some point, the MCO was going to warn folks who had outdated kubelets that they were behind, with hints about how to update.  Or maybe they were going to warn folks who were up against the boundary (so with the Kube-API-server's `Upgradeable=False` going off), with hints about how to update.  Staying silent until the nodes are _past_ the boundary seems like it's waiting too long.

[1]: https://github.com/openshift/enhancements/blame/3716ed25593447d9e624cb4fbabd90fcec66c58f/enhancements/update/eus-upgrades-mvp.md
[2]: https://github.com/openshift/enhancements/pull/800/files#diff-91c1b36fc770bc12ef9147ff77360aa00fb1f79d3df986a471d75d1c4a33891cR292-R293

kikisdeliveryservice replied:
This goes hand in hand with kube-apiserver, but might be worth having 2 cases:

1) KubeletSkewUnsupported for people past the boundary (which I think can happen since this won't be backported to infinity?) 
2) KubeletSkewPresent (or some better name than what I can think of right now) where we are informationally saying: that we recommend upgrading your pool (maybe they are just 1 skew), let them know then and maybe they will just upgrade and save us all the trouble? 

Basically if there is skew you always say something, but you'll probably want to say different things

QiWang19 replied:
@kikisdeliveryservice Follow your suggestion, what if change to the following check and warn the users there's going to be an unsupported version skew?
```go
minorVSkew, kasV, kubeletV := optr.isKubeletSkewSupported()
if minorVSkew > 2 {
	coStatus.Reason = "KubeletSkewUnsupported"
	coStatus.Message = fmt.Sprintf("one or more nodes have an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least %v.", getMinimalSkewSupportNodeVersion(v))
} else if minorVSkew == 2 {
	coStatus.Reason = "KubeletSkewPresent"
	coStatus.Message = fmt.Sprintf("current kubelet version %v will not be supported by newer kube-apiserver. Please upgrade the the kubelet first if plan to upgrade the kube-apiserver")
}
```

kikisdeliveryservice replied:
Skew is also present if the diff is 1 and I think Trevor is pointing out that we should say something whenever there is a skew and not wait.

rphillips replied:
There will be a skew regardless on upgrades. Is the intent to report to the user on an upgrade there is a skew?

rphillips replied:
The enhancement is for >=2 skew... If we want to create a further enhancement for 4.10 for skew < 2 we can do that. It's out of scope of this PR.

QiWang19 commented:
/retest

QiWang19 commented:
@kikisdeliveryservice PTAL. I saw the KubeletSkewUnsupported on the cluster, apiserver version is 1.21.1, kubelet version is 1.18.0. I will clean up new reviews and get must-gather.
```
$ os describe ClusterOperator
    Last Transition Time:  2021-07-20T22:28:30Z
    Message:               machine config pool contains nodes with an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet Version of at least 1.19.1.
    Reason:                KubeletSkewUnsupported
    Status:                True
    Type:                  Upgradeable
```

QiWang19 commented:
In the must-gather, the event shows in `registry-build01-ci-openshift-org-ci-ln-m0wnn6k-stable-sha256-955070604a96dd31ae3e857876595d97fc13eea864445a89abfa2f634d1929fa/namespaces/openshift-machine-config-operator/pods/machine-config-operator-57fcc46c4d-27xl7/machine-config-operator/machine-config-operator/logs/current.log`
[must-gather.tar.gz](https://github.com/openshift/machine-config-operator/files/6856568/must-gather.tar.gz)

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +291,105 @@ func (optr *Operator) syncUpgradeableStatus() error {
+// Returns false if the version skew is not supported. version skew > 2 is not supported.
+func (optr *Operator) isKubeletSkewSupported() (skewSupported bool, kubeAPIServerVersion string) {
+	skewSupported = true
+	nodes, err := optr.nodeLister.List(labels.Everything())
```
One question here: will this pull ALL the nodes in the cluster? bc after thinking about this we only want it to pull the nodes managed by the MCO (ie nodes that are in machine config pools)?

rphillips replied:
Great question. I think we want all the nodes in the cluster, since a version skew that goes out of bounds might leave a kubelet not able to communicate to the API server.

kikisdeliveryservice replied:
The problem with this is that those nodes wouldn't be managed by the MCO and by reporting it in our status we seem to be implying that we, in fact, do.

kikisdeliveryservice replied:
The more I consider this, the less it makes sense to me as it's not within the operator's mandate to consider and/or touch nodes that are not members of MCPs.

QiWang19 replied:
If we don't check the non MCO nodes, not sure if other components will do the check?
What will be the result of that wrong implication? I saw the MCO manage the nodes using labels, if they use the MCO to manage the node, will they get notified of missing labels? So it will not be a big concern here.

kikisdeliveryservice replied:
Back to my original question: does this check pull in non-MCO managed nodes? If it does, the concern is that the MCO only manages nodes that are members of a pool, we do not manage, upgrade, etc.. nodes that are not in pools. Verifying/managing  these non-pool nodes is very much out of scope for the operator. Reporting their status as part of our operator status creates an untrue implication that the MCO manages them/watches them generally.

kikisdeliveryservice replied:
Users are generally responsible for non-MCO managed nodes in a cluster, I don't think we want that relationship to change.

QiWang19 replied:
Fixed.

QiWang19 commented:
@kikisdeliveryservice @sinnykumari PTAL. Added a commit https://github.com/openshift/machine-config-operator/pull/2658/commits/07d50206435b7ab99683d9878af255e20b1aed91 for the above reviews about the coStatus.Message and log errors.

kikisdeliveryservice commented:
> @kikisdeliveryservice @sinnykumari PTAL. Added a commit [07d5020](https://github.com/openshift/machine-config-operator/commit/07d50206435b7ab99683d9878af255e20b1aed91) for the above reviews about the coStatus.Message and log errors.

This is failing verify and unity tests. Could you please fix before we review?

QiWang19 commented:
> > @kikisdeliveryservice @sinnykumari PTAL. Added a commit [07d5020](https://github.com/openshift/machine-config-operator/commit/07d50206435b7ab99683d9878af255e20b1aed91) for the above reviews about the coStatus.Message and log errors.
> 
> This is failing verify and unit tests. Could you please fix before we review?

Fixed. PTAL

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +301,165 @@ func (optr *Operator) syncUpgradeableStatus() error {
+		coStatus.Message = fmt.Sprintf("An error occurred when checking kubelet version skew: %v", err)
+		return skewUnchecked, coStatus, err
+	}
+	workerNodes, err := optr.nodeLister.List(labels.SelectorFromSet(labels.Set{node.WorkerLabel: ""}))
```
Clusters can have custom pools as well.

rphillips replied:
They can have custom pools, but they are always Masters or Workers, right?

kikisdeliveryservice replied:
custom pools don't have to have a worker label: https://github.com/openshift/machine-config-operator/blob/master/docs/custom-pools.md#creating-a-custom-pool

QiWang19 replied:
Does those nodes have label llike `labels.SelectorFromSet(labels.Set{fmt.Sprintf("node-role.kubernetes.io/%s", pool.Name): ""})`?
If they do we can list the pools and filter the nodes by the label.

QiWang19 replied:
@kikisdeliveryservice  As https://github.com/openshift/machine-config-operator/pull/2658#issuecomment-885218200 commeted, the new function will get the mco managed node.

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +301,165 @@ func (optr *Operator) syncUpgradeableStatus() error {
+			return skewUnsupported, coStatus, nil
+		}
+		if nodeMinorVersion+2 == kubeAPIServerMinorVersion {
+			coStatus.Reason = "KubeletSkewPresent"
```
why is this not the value of the const?

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -243,6 +246,15 @@ func (optr *Operator) syncDegradedStatus(ierr syncError) (err error) {
+	skewSupported
+	skewUnsupported
+	skewPresent
+)
```
Can you explain what's happening here with these const? Why not assign them the value of the Reason strings?

kikisdeliveryservice reviewed (changes requested):
verify is still failing + some below comments since some of this seems confusing.

@rphillips there have been a bunch of changes, can you review/verify and let us know when this is ready?

QiWang19 commented:
The new commit is a tentative commit based on the comments about report error if the version skew has been checked  https://github.com/openshift/machine-config-operator/pull/2658#discussion_r673590213 and about the KubeletSkewPresent https://github.com/openshift/machine-config-operator/pull/2658#discussion_r673567119

rphillips commented:
We've add a few fixes to this PR from feedback here. A log line for the operator was added [here](https://github.com/openshift/machine-config-operator/pull/2658/files#diff-8a6a064b4db19eb3b29e5280447ac3ba9496e847391489d25c1282767c3354c5R310). Additionally, we added a new function `GetAllManagedNodes` to gather all the managed nodes in the system from all the pools. This will cover custom pools and the default pools we create.

rphillips commented:
@kikisdeliveryservice I believe this PR is ready... Takes into account all discussions on this thread.

sinnykumari commented on pkg/operator/status.go:
```diff
@@ -289,9 +299,172 @@ func (optr *Operator) syncUpgradeableStatus() error {
+		}
+		if nodeMinorVersion+2 == kubeAPIServerMinorVersion {
+			coStatus.Reason = skewPresent
+			coStatus.Message = fmt.Sprintf("Current kubelet version %v will not be supported by newer kube-apiserver. Please upgrade the the kubelet first if plan to upgrade the kube-apiserver", kubeletVersion)
```
```suggestion
			coStatus.Message = fmt.Sprintf("Current kubelet version %v will not be supported by newer kube-apiserver. Please upgrade the kubelet first if plan to upgrade the kube-apiserver", kubeletVersion)
```

sinnykumari commented on pkg/operator/status.go:
```diff
@@ -289,9 +299,172 @@ func (optr *Operator) syncUpgradeableStatus() error {
+			coStatus.Reason = skewPresent
+			coStatus.Message = fmt.Sprintf("Current kubelet version %v will not be supported by newer kube-apiserver. Please upgrade the the kubelet first if plan to upgrade the kube-apiserver", kubeletVersion)
+			return skewPresent, coStatus, nil
+		}
```
Looks like for `nodeMinorVersion == kubeAPIServerMinorVersion`, this function will return skewSupported which couldn't be confusing sine there is no skew at all. Shouldn't we return `noSkew` for this case?

QiWang19 replied:
renamed `skewSupported` to `noSkew`.

QiWang19 replied:
Are you suggesting `nodeMinorVersion == kubeAPIServerMinorVersion` shouldbe `noSkew` and `kubeAPIServerMinorVersion - nodeMinorVersion == 1` shoule be `skewSupported`?

For `coStatus`, `kubeAPIServerMinorVersion - nodeMinorVersion < 2` is out of the scoop of this PR https://github.com/openshift/machine-config-operator/pull/2658#discussion_r675082603, I don't see it is necessary for `isKubeletSkewSupported` to descrimintate it at this moment.

rphillips replied:
@sinnykumari I think you misread the line perhaps? there is a +2 which would be skewPresent.

sinnykumari replied:
What I meant was to add a new field `noSkew` to handle condition `nodeMinorVersion == kubeAPIServerMinorVersion`. If you replace it, that would mean for `nodeMinorVersion+1 == kubeAPIServerMinorVersion` as well it will return noSkew, although there is skewPresent. I think @rphillips  can tell better what is best here. Because in the end how these status is being consumed and message we want to give customer matters.

QiWang19 replied:
I revert the rename change, keep using `skewSupported` for `kubeAPIServerMinorVersion - nodeMinorVersion < 2`  until a consensus is reached.

kikisdeliveryservice replied:
@sinnykumari is this now ok?

sinnykumari replied:
yeah, based on https://github.com/openshift/machine-config-operator/pull/2658#discussion_r675082603 it seems handling n<2 is not part of current enhancement. So, it is upto node team how they get handled in 4.9.

sinnykumari commented on pkg/operator/status.go:
```diff
@@ -289,9 +299,172 @@ func (optr *Operator) syncUpgradeableStatus() error {
+		if err != nil {
+			coStatus.Reason = status.Reason
+			coStatus.Message = status.Message
+			return optr.updateStatus(co, coStatus)
```
Can we log error message here as well so that operator pod log captures it?

QiWang19 replied:
Added glof.Fatalf

rphillips replied:
we don't want Fatalf, that will cause the controller to exit.

QiWang19 replied:
fixed.

sinnykumari commented on pkg/operator/status.go:
```diff
@@ -289,9 +299,172 @@ func (optr *Operator) syncUpgradeableStatus() error {
+			coStatus.Message = status.Message
+			return optr.updateStatus(co, coStatus)
+		}
+		glog.Infof("kubelet skew status: %v, status reason: %v, status message: %v", skewStatus, status.Reason, status.Message)
```
I think we should not log message when there is no version skew, otherwise operator log will be flooded with the message while user doesn't have to take any action. Related to https://github.com/openshift/machine-config-operator/pull/2658/files#r675479794 . We may have to do same for `skewSupported` as well but that can be done later on as well.

sinnykumari commented:
Thanks for making all the changes! Few minor suggestions, other than that LGTM

QiWang19 commented:
@sinnykumari PTAL.

sinnykumari commented:
I will let Kirsten do final review.

QiWang19 commented:
@kikisdeliveryservice PTAL.

kikisdeliveryservice reviewed (commented):
A few comments.

Can we also see a must gather since there've been so many changes to this PR?

Thanks!

QiWang19 commented:
> A few comments.
> 
> Can we also see a must gather since there've been so many changes to this PR?
> 
> Thanks!

[must-gather.tar.gz](https://github.com/openshift/machine-config-operator/files/6885819/must-gather.tar.gz)
must-gather attached.  apiserver version is 1.21.1, kubelet version is 1.18.0.
```
// namespaces/openshift-machine-config-operator/pods/machine-config-operator-bbb5555d6-z5rvz/machine-config-operator/machine-config-operator/logs/current.log
2021-07-27T03:55:19.559426141Z I0727 03:55:19.559244       1 start.go:43] Version: 4.9.0-0.ci.test-2021-07-27-033800-ci-ln-04clxvb-latest (Raw: machine-config-daemon-4.6.0-202006240615.p0-975-g9f036fe9-dirty, Hash: 9f036fe9b49b6de2eacc991c37678587e14c0386)
2021-07-27T03:55:19.562886709Z I0727 03:55:19.562779       1 leaderelection.go:243] attempting to acquire leader lease openshift-machine-config-operator/machine-config...
2021-07-27T03:57:15.259786604Z I0727 03:57:15.259685       1 leaderelection.go:253] successfully acquired lease openshift-machine-config-operator/machine-config
2021-07-27T03:57:16.311510738Z I0727 03:57:16.311439       1 operator.go:262] Starting MachineConfigOperator
2021-07-27T03:57:22.603593629Z I0727 03:57:22.603512       1 status.go:325] kubelet skew status: KubeletSkewUnsupported, status reason: KubeletSkewUnsupported
2021-07-27T03:57:22.603811222Z I0727 03:57:22.603689       1 event.go:282] Event(v1.ObjectReference{Kind:"", Namespace:"", Name:"machine-config", UID:"4753add8-791d-4b39-8ab8-3ecfbc169be2", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'KubeletSkewUnsupported' One or more nodes have an unsupported kubelet version skew. Please see `oc get nodes` for details and upgrade all nodes so that they have a kubelet version of at least 1.19.1.
```

QiWang19 commented on pkg/operator/status.go:
```diff
@@ -289,9 +299,172 @@ func (optr *Operator) syncUpgradeableStatus() error {
+				Namespace: co.Namespace,
+				UID:       co.GetUID(),
+			}
+			optr.eventRecorder.Eventf(mcoObjectRef, corev1.EventTypeWarning, coStatus.Reason, fmt.Sprintf("machine config pool contains nodes with an unsupported kubelet version skew."))
```
Fixed

QiWang19 commented on pkg/operator/status.go:
```diff
@@ -289,9 +299,172 @@ func (optr *Operator) syncUpgradeableStatus() error {
+				Namespace: co.Namespace,
+				UID:       co.GetUID(),
+			}
+			optr.eventRecorder.Eventf(mcoObjectRef, corev1.EventTypeWarning, coStatus.Reason, fmt.Sprintf("machine config pool contains nodes with an N-1 kubelet version skew."))
```
Fixed

QiWang19 commented:
/bugzilla refresh

QiWang19 commented on pkg/operator/status.go:
```diff
@@ -289,9 +299,175 @@ func (optr *Operator) syncUpgradeableStatus() error {
+	if !updating && !degraded {
+		skewStatus, status, err := optr.isKubeletSkewSupported(pools)
+		if err != nil {
+			glog.Errorf("Error checking version skew: %v, kubelet skew status: %v, status reason: %v, status message: %v", err, skewStatus, status.Reason, status.Message)
```
Wheather this `glog.Errorf` casued the CI failures? https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/2658/pull-ci-openshift-machine-config-operator-master-okd-e2e-aws/1419855633997369344/artifacts/e2e-aws/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-operator-598979445f-sd8qf_machine-config-operator.log

kikisdeliveryservice commented:
/test e2e-agnostic-upgrade

rphillips commented:
I tested this PR locally and it does report the skew, and go back to no skew as expected.

kikisdeliveryservice commented on pkg/operator/status.go:
```diff
@@ -289,9 +299,175 @@ func (optr *Operator) syncUpgradeableStatus() error {
+				UID:       co.GetUID(),
+			}
+			glog.Infof("kubelet skew status: %v, status reason: %v", skewStatus, status.Reason)
+			optr.eventRecorder.Eventf(mcoObjectRef, corev1.EventTypeWarning, coStatus.Reason, coStatus.Message)
```
just talked to ryan, and @QiWang19 can you remove this event? Since this is a supported state that might last fora  while, we don't think that filling the MCO's logs with it every 5 mins for a long time is going to be helpful and will create v v large log files.

QiWang19 replied:
Removed.

kikisdeliveryservice replied:
thanks Qi!

kikisdeliveryservice reviewed (commented):
one last thing otherwise looks good

kikisdeliveryservice reviewed (approved):
Thanks for all of the work on this. I think we've gotten it to a good state.

rphillips commented:
/lgtm
